### ISSUE TITLE
How to attach custom vertex information in smooth terrain?

### ISSUE THREAD

**Issue by id05** [Zylann/godot_voxel#590]
How to attach custom data to vertexes in transvoxel mesher? Color, for example... 

I tried to figure it out myself, but I know C++ only on basic level

**Comment by Zylann** [comment]
You can't attach custom data to meshes without modifying the mesher in C++. Also, that data needs to come from somewhere, and usually it's voxel data (one of the channels). 
That data can be attached either during polygonization (somewhere inside [regular mesh](https://github.com/Zylann/godot_voxel/blob/affbadeb1ab456d1e72a54dd9090ce0d2d5f8f2c/meshers/transvoxel/transvoxel.cpp#L315) and [transition mesh](https://github.com/Zylann/godot_voxel/blob/affbadeb1ab456d1e72a54dd9090ce0d2d5f8f2c/meshers/transvoxel/transvoxel.cpp#L1458) functions), or afterwards when vertex data is [converted](https://github.com/Zylann/godot_voxel/blob/affbadeb1ab456d1e72a54dd9090ce0d2d5f8f2c/meshers/transvoxel/voxel_mesher_transvoxel.cpp#L78) to Godot's format. In addition, just "adding" the data from voxel data isn't always trivial depending on what you want to do, because vertices can be anywhere between voxel cells. I suppose you could interpolate voxel colors around each vertex and add it to the COLOR attribute (currently unused).
Note 1, there is already support for texturing data.
Note 2, transvoxel meshes are relatively heavy, so adding more data to meshes may impact rendering performance a bit (in addition to CPU work for attaching the data).

**Comment by id05** [comment]
Yes, the idea was to tinting the meshes, to make an appearance of different materials. 
I was reading #560 and I thought for a second, that make same effect with colors and 16 textures would be more simple, then changing texturing logic

**Comment by Zylann** [comment]
If it's just a visual trick to fake diversity of textures, using voxels and adding extra mesh data is probably not a good idea. I would rather use a shader trick for doing this, like using a few layers of noise textures.
